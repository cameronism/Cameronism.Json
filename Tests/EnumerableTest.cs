using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using UnsafeJson;
using Xunit;

namespace Tests
{
	public class EnumerableTest
	{
		[Fact]
		public void FindMethods()
		{
			var sb = new StringBuilder();

			foreach (var type in new[]
			{
				typeof(List<int>),
				typeof(IEnumerable<int>),
			})
			{
				var methods = EnumerableInfo.FindMethods(type);
				sb.AppendLine("# " + SchemaTest.HumanName(type));
				DescribeMethod(sb, () => methods.GetEnumerator);
				DescribeMethod(sb, () => methods.MoveNext);
				DescribeMethod(sb, () => methods.get_Current);
				DescribeMethod(sb, () => methods.get_Count);
				DescribeMethod(sb, () => methods.Dispose);

				sb.AppendLine();
			}

			ApprovalTests.Approvals.Verify(sb.ToString());
		}

		static void DescribeMethod(StringBuilder sb, Expression<Func<MethodInfo>> fmi)
		{
			var prop = fmi.Body as MemberExpression;
			sb.AppendLine("## " + prop.Member.Name);

			var mi = fmi.Compile().Invoke();
			if (mi == null)
			{
				sb.AppendLine("`null`");
			}
			else
			{
				sb.AppendLine("Return Type: " + SchemaTest.HumanName(mi.ReturnType));
				sb.AppendLine("IsVirtual: " + mi.IsVirtual);
				sb.AppendLine("Declaring Type: " + SchemaTest.HumanName(mi.DeclaringType));
				foreach (var arg in mi.GetParameters())
				{
					sb.AppendLine(String.Format("{0}: {1}", arg.Name, SchemaTest.HumanName(arg.ParameterType)));
				}
			}

			sb.AppendLine();
		}



	}
}
