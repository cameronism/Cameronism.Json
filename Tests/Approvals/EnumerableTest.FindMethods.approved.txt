fixme still need to get more selective about which methods to use

# System.Collections.Generic.List<System.Int32>
## GetEnumerator
Return Type: System.Collections.Generic.List<System.Int32>+Enumerator
IsVirtual: False
Declaring Type: System.Collections.Generic.List<System.Int32>

## MoveNext
Return Type: System.Boolean
IsVirtual: True
Declaring Type: System.Collections.Generic.List<System.Int32>+Enumerator

## get_Current
Return Type: System.Int32
IsVirtual: True
Declaring Type: System.Collections.Generic.List<System.Int32>+Enumerator

## get_Count
Return Type: System.Int32
IsVirtual: True
Declaring Type: System.Collections.Generic.List<System.Int32>

## Dispose
Return Type: System.Void
IsVirtual: True
Declaring Type: System.Collections.Generic.List<System.Int32>+Enumerator


# System.Collections.Generic.IEnumerable<System.Int32>
## GetEnumerator
Return Type: System.Collections.Generic.IEnumerator<System.Int32>
IsVirtual: True
Declaring Type: System.Collections.Generic.IEnumerable<System.Int32>

## MoveNext
Return Type: System.Boolean
IsVirtual: True
Declaring Type: System.Collections.IEnumerator

## get_Current
Return Type: System.Int32
IsVirtual: True
Declaring Type: System.Collections.Generic.IEnumerator<System.Int32>

## get_Count
`null`

## Dispose
Return Type: System.Void
IsVirtual: True
Declaring Type: System.IDisposable


